### DISCUSSION TITLE
tar_group data.frame helper?

### DISCUSSION THREAD

**Discussion by liutiming** [ropensci/targets#329] [Ideas]
I was wondering whether a function like the following would be helpful to many people. I can create a PR if so...

```
#' group a table into n tar_group for use by [targets::tar_group()]
id_group  <- function(id_tbl, n) {
    id_tbl  %>% 
    rowid_to_column  %>% 
    mutate(group_num = rowid %% n + 1) %>% group_by(group_num) %>% tar_group() 
}
```

**Comment by wlandau** [comment]
Is the idea to treat each row as a group? If so, `iteration = "vector"` already covers this, thanks to `vctrs` magic.

``` r
library(targets)
tar_script({
  library(gapminder)
  list(
    tar_target(data, gapminder),
    tar_target(each_row, data, pattern = map(data))
  )
})

tar_make(reporter = "silent")

tar_read(each_row, branches = 1)
#> # A tibble: 1 x 6
#>   country     continent  year lifeExp     pop gdpPercap
#>   <fct>       <fct>     <int>   <dbl>   <int>     <dbl>
#> 1 Afghanistan Asia       1952    28.8 8425333      779.

tar_read(each_row, branches = 2)
#> # A tibble: 1 x 6
#>   country     continent  year lifeExp     pop gdpPercap
#>   <fct>       <fct>     <int>   <dbl>   <int>     <dbl>
#> 1 Afghanistan Asia       1957    30.3 9240934      821.
```

<sup>Created on 2021-02-26 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by liutiming** [comment]
thanks! not each row, but per n rows -> 1 group

**Reply by wlandau** [reply]
Interesting. Maybe this would be a good addition to `tarchetypes`. `tar_group_id(name, produce_df(), rows = 10)` could be equivalent to:

```r
tar_target(
  name,
  { # best to put this in an exported tarchetypes function with `@keywords internal`
    out <- produce_df()
    out$tar_group <- seq_len(nrow(out)) %% rows + 1L # does the work of tar_group()
    out
  },
  format = "fst_tbl" # default format but could be modified
  iteration = "group" # user should not be able to modify
)
```

We might also consider a `tar_group_by(name, produce_df(), by = c("var1", "var2"))` helper equivalent to:

```r
tar_target(
  name, {
    produce_df() %>%
      group_by(var1, var2) %>%
      tar_group() %>%
      ungroup()
  },
  format = "fst_tbl"
  iteration = "group"
)
```

**Reply by wlandau** [reply]
Do you have a strong interest in implementation? Otherwise, I will. I feel these are important helpers.

**Reply by wlandau** [reply]
On reflection, I now have strong opinions on what these functions should be named and how they should be implemented, so I would prefer to implement them myself.

Your idea is super nice, and I think it will make grouped operations a lot easier.

**Reply by liutiming** [reply]
aha thanks! sure I look forward to it!
